packages/postcss-tape/src/index.ts

```diff
@@ -0,0 +1,344 @@
+						console.log(formatGitHubActionAnnotation(
+							`${testCaseLabel}\n\nbroken source map: ${JSON.stringify(result.map.toJSON().sources)}`,
+							'error',
+							{ file: testFilePath, line: 1, col: 1 },
+						));
+					} else {
+						console.error(`\n${testCaseLabel}\n\nbroken source map: ${JSON.stringify(result.map.toJSON().sources)}\n\n${dashesSeparator}`);
+					}
+				}
+			}
+
+			// Run "after" when initial postcss run has completely.
+			if (testCaseOptions.after) {
+				await testCaseOptions.after();
+			}
+
+			// Assert that the result can be passed back to PostCSS and still parses.
+			{
+				try {
+					const resultContents = await fsp.readFile(resultFilePath, 'utf8');
+					const secondPassResult = await postcss([noopPlugin()]).process(resultContents, {
+						from: resultFilePath,
+						to: resultFilePath,
+						map: {
+							inline: false,
+							annotation: false,
+						},
+					});
+
+					if (secondPassResult.warnings().length) {
+						throw new Error('Unexpected warnings on second pass');
+					}
+				} catch (_) {
+					hasErrors = true;
+
+					if (process.env.GITHUB_ACTIONS) {
+						console.log(formatGitHubActionAnnotation(
+							`${testCaseLabel}\n\nresult was not parsable with PostCSS.`,
+							'error',
+							{ file: expectFilePath, line: 1, col: 1 },
+						));
+					} else {
+						console.error(`\n${testCaseLabel}\n\nresult was not parsable with PostCSS.\n\n${dashesSeparator}`);
+					}
+				}
+			}
+
+			// Assert result with oldest supported PostCSS.
+			// Check against the actual result to avoid duplicate warnings.
+			//
+			// NOTE:
+			// The oldest version of PostCSS we support at this time.
+			// This does not need to be a different version than the latest version of PostCSS.
+			// There is no system behind setting this.
+			// It is here to allow testing a specific older version, if parts of the community can't update yet.
+			{
+				const resultFromOldestPostCSS = await postcssOldestSupported(plugins as Array<AcceptedPlugin>).process(input, {
+					from: testFilePath,
+					to: resultFilePath,
+					map: {
+						inline: false,
+						annotation: false,
+					},
+				});
+
+				try {
+					assert.strictEqual(resultFromOldestPostCSS.css.toString(), resultString);
+				} catch (err) {
+					hasErrors = true;
+
+					if (process.env.GITHUB_ACTIONS) {
+						console.log(formatGitHubActionAnnotation(
+							'testing older PostCSS:\n' + formatCSSAssertError(testCaseLabel, testCaseOptions, err, true),
+							'error',
+							{ file: normalizeFilePathForGithubAnnotation(expectFilePath), line: 1, col: 1 },
+						));
+					} else {
+						console.error('testing older PostCSS:\n' + formatCSSAssertError(testCaseLabel, testCaseOptions, err));
+					}
+				}
+			}
+
+			// Assert that warnings have the expected amount.
+			{
+				try {
+					if (result.warnings().length || testCaseOptions.warnings) {
+						assert.strictEqual(result.warnings().length, testCaseOptions.warnings);
+					}
```

Comment:
Should this be averaged between all of the runs so we make sure it's not warning anywhere else?

Reply:
Do you mean averaging between possible older en current postcss versions?

-------

_if not, you can ignore the bit here_

I have been trying to think of a case were this could differ and I can only think of things that would be intentional. (warning users that they should update postcss because feature X is required or gives a better result)

The test against older postcss is intended to catch exceptions that are thrown when using newer API's.

Reply:
Yeah, actually thought the same, just wanted to ponder the question. I agree that's the only usecase that comes to mind, throwing warnings for deprecations, etc.